docs/dev/relay_bring_your_own_codegen.rst

```diff
@@ -0,0 +1,498 @@
+
+.. code-block:: c++
+
+  runtime::Module CreateCSourceModule(const NodeRef& ref) override {
+    // Create headers
+    code_stream_ << "#include <cstdint>\n";
+    code_stream_ << "#include <iostream>\n";
+    code_stream_ << "#include <cstdlib>\n";
+    code_stream_ << "#include <stdio.h>\n";
+    code_stream_ << "#include <cstring>\n";
+    code_stream_ << "#include <tvm/runtime/c_runtime_api.h>\n";
+    code_stream_ << "#include <dlpack/dlpack.h>\n";
+
+    // Append some common macro for operator definition.
+    const char* operator_macro = R"op_macro(
+    #define CSOURCE_BINARY_OP_1D(p_ID_, p_OP_, p_DIM1_)       \
+      extern "C" void p_ID_(float* a, float* b, float* out) { \
+        for (int64_t i = 0; i < p_DIM1_; ++i) {               \
+          out[i] = a[i] p_OP_ b[i];                           \
+        }                                                     \
+      }
+    
+    #define CSOURCE_BINARY_OP_2D(p_ID_, p_OP_, p_DIM1_, p_DIM2_)  \
+      extern "C" void p_ID_(float* a, float* b, float* out) {     \
+        for (int64_t i = 0; i < p_DIM1_; ++i) {                   \
+          for (int64_t j = 0; j < p_DIM2_; ++j) {                 \
+            int64_t k = i * p_DIM2_ + j;                          \
+            out[k] = a[k] p_OP_ b[k];                             \
+          }                                                       \
+        }                                                         \
+      }
+    )op_macro";
+
+    code_stream_ << operator_macro << "\n\n";
+
+    // Generate C code for the subgraph.
+    if (ref->IsInstance<FunctionNode>()) {
+      GenCFunc(Downcast<Function>(ref));
+    } else if (ref->IsInstance<relay::ModuleNode>()) {
+      relay::Module mod = Downcast<relay::Module>(ref);
+      for (const auto& it : mod->functions) {
+        GenCFunc(Downcast<Function>(it.second));
+      }
+    } else {
+      LOG(FATAL) << "The input ref is expected to be a Relay function or module"
+                 << "\n";
+    }
+
+    // Create a CSourceModule
+    const auto* pf = runtime::Registry::Get("module.csource_module_create");
+    CHECK(pf != nullptr) << "Cannot find csource module to create the external runtime module";
+    return (*pf)(code_stream_.str(), "cc");
+  }
+
+Register Your Codegen
+=====================
+
+The last step is registering your codegen to TVM backend. We first implement a simple function to invoke our codegen and generate a runtime module.
+
+.. code-block:: c++
+
+  runtime::Module CCompiler(const NodeRef& ref) {
+    CSourceCodegen csource;
+    return csource.CreateCSourceModule(ref);
+  }
+
+Finally, we register this function to TVM backend:
+
+.. code-block:: c++
+
+  TVM_REGISTER_API("relay.ext.ccompiler").set_body_typed(CCompiler);
+
+where ``ccompiler`` is a customized tag to let TVM know this is the codegen it should use to generate and offload subgraphs when the subgraph is annotated with ``ccompiler``.
```

Comment:
which means the generated code could be compiled by any C/C++ compiler with proper compilation flags.

Reply:
This sentence looks weird here so I put it at the begining to explain the benefit of generating C code.